src/gpio/wait_asynch.rs

```diff
@@ -0,0 +1,110 @@
+use core::{
+    future::Future,
+    sync::atomic::{AtomicBool, Ordering},
+    task::Poll,
+};
+extern crate alloc;
+use crate::gpio::{InputMode, InterruptType, Pin, PinDriver};
+use alloc::sync::Arc;
+use atomic_waker::AtomicWaker;
+use esp_idf_sys::{esp_nofail, gpio_intr_disable, EspError};
+
+impl<T: Pin, MODE: InputMode> embedded_hal_async::digital::Wait for PinDriver<'_, T, MODE> {
+    async fn wait_for_high(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::HighLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_low(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::LowLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_rising_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::PosEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_falling_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::NegEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_any_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::AnyEdge)?.await;
+        Ok(())
+    }
+}
+
+pub(crate) struct InputFuture<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode> {
+    // Unfortunately, the Wait trait uses functions that are given a mutable
+    // reference to the pin driver with no explicit lifetime parameter.  Since
+    // the reference has a different lifetime than the struct, we must make this
+    // explicit, or else the borrow checker will believe the reference must live
+    // forever.
+    driver: &'driver_ref mut PinDriver<'driver_struct, T, MODE>,
+    processed_interrupt: Arc<AtomicBool>,
+    atomic_waker: Arc<AtomicWaker>,
+}
+
+impl<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode>
+    InputFuture<'driver_ref, 'driver_struct, T, MODE>
+{
+    pub(crate) fn new(
+        driver: &'driver_ref mut PinDriver<'driver_struct, T, MODE>,
```

Comment:
Orthogonal note: a lot of what you do here (but not everything) can be perhaps simplified by reusing [this](https://github.com/ivmarkov/channel-bridge/blob/master/src/notification.rs#L21) or [this](https://github.com/embassy-rs/embassy/blob/master/embassy-sync/src/signal.rs#L34). Should we depend on one of these instead? Perhaps the first one, given your inclination towards lock-free.

Reply:
Happy to do it, as per the embassy links above, i just used a familiar solution :)

Reply:
You can use embassy's `Signal` just fine from within an ISR context [with this guy](https://github.com/esp-rs/esp-idf-hal/blob/master/src/interrupt.rs#L241). One of the reasons why I parameterized `Signal` by the mutex type in the first place.

Reply:
Okay. Do you want me to move it to use signals rather than notifications?
I'm happy to do whatever.

Reply:
You can stick to notifications. This comment was "FYI" just to address your comments that most forms of locks cannot be used in an ISR. ISR critical sections can - that's all.